collector/zfs_freebsd.go

```diff
@@ -0,0 +1,115 @@
+package collector
+
+import (
+	"bufio"
+	"io"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+/*
+#cgo LDFLAGS:
+#include <sys/param.h>
+#include <sys/module.h>
+
+int zfsModuleLoaded() {
+	int modid = modfind("zfs");
+	return modid < 0 ? 0 : -1;
+}
+
+*/
+import "C"
+
+func (c *zfsCollector) zfsAvailable() error {
+	if C.zfsModuleLoaded() == 0 {
+		return zfsNotAvailableError
+	}
+	return nil
+}
+
+const zfsArcstatsSysctl = "kstat.zfs.misc.arcstats"
+
+func (c *zfsCollector) RunOnStdout(cmd *exec.Cmd, handler func(io.Reader) error) (err error) {
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return
+	}
+
+	if err = cmd.Start(); err != nil {
+		return
+	}
+
+	err = handler(stdout)
+	if err != nil {
+		return
+	}
+
+	if err = cmd.Wait(); err != nil {
+		return
+	}
+
+	return err
+
+}
+
+func (c *zfsCollector) updateArcstats(ch chan<- prometheus.Metric) (err error) {
+
+	cmd := exec.Command("sysctl", zfsArcstatsSysctl)
```

Comment:
Originally, I used the sysctl C API which is more than a little painful to use when you don't know the exact list of sysctl's to query for.
In the ZoL part, we simply expose all variables in the file.
I figured this would be a good idea for FreeBSD (not baking in a fixed set of sysctls).

FreeBSD only has rudimentary and little/non-supported `procfs` for Linux-compatibility. It is not mounted by default.

I tried to extract the 'discover all available sysctls'-part from [FreeBSD's sysctl implementation](https://github.com/freebsd/freebsd/blob/master/sbin/sysctl/sysctl.c) but wasn't successful in finding any documentation about the way they do it there.

I don't like building against undocumented / obscure behavior since it may break in future OS releases.

If a C implementation on FreeBSD is desired I will look into it again. However, because of

* simplicity
* maintainability
* no drastic performance improvement (sysctl(8) does way more syscalls than a simple fork/exec for discovery of all sysctls anyways)

I strongly advise against it.